### PR TITLE
Module 3 Project Exercise

### PR DESCRIPTION
**Pull request by shivaHanchate**
Hi,
Can you please review my module 3 project exercise and let me know if i need to make any changes.

Many Thanks
Shiva

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -116,4 +116,11 @@ env/
...
-config.py
\ No newline at end of file
+config.py
+
+### geckodriver.exe and geckodriver.log ###
+geckodriver.exe
+geckodriver.log
+
+.exe
+.log
```
Comment (hugh-emerson) [inline .gitignore]: I think you meant `*.exe` and `*.log`
Slightly odd to blanket ignore all exe files though.

File: .idea/DevOps-Course-Starter.iml [hunk 2]
```diff
@@ -2,7 +2,7 @@
 <module type="PYTHON_MODULE" version="4">
   <component name="NewModuleRootManager">
     <content url="file://$MODULE_DIR$" />
-    <orderEntry type="inheritedJdk" />
+    <orderEntry type="jdk" jdkName="Python 3.8 (DevOps-Course-Starter)" jdkType="Python SDK" />
```
Comment (hugh-emerson) [inline .idea/DevOps-Course-Starter.iml:5]: You have both .idea and .vscode files - are you switching between two IDEs? We would recommend sticking to VSCode for this course

File: README.md [hunk 3]
```diff
@@ -38,6 +38,17 @@ Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser
 
 The `.env` file is used by flask to set environment variables when running `flask run`. This enables things like developement mode (which also enables features like hot reloading when you make a file change).
 
-please add API key, token, board_id, todo_list_id, doing_list_id and done_list_id variables to .env to match your Trello details.
+Please add API key, token, board_id, todo_list_id, doing_list_id and done_list_id variables to .env to match your Trello details.
+
+Please add API key, token, board_id, todo_list_id, doing_list_id and done_list_id variables to .env.test beforerunning the test.
```
Comment (hugh-emerson) [inline README.md]: No, .env.test should only be used by unit/integration tests and should not contain real values. So the version you've committed shouldn't need to be touched

File: README.md [hunk 4]
```diff
@@ -38,6 +38,17 @@ Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser
...
+
+Please download and copy geckodriver.exe into DEVOPS-COURSE-STARTER folder.
```
Comment (hugh-emerson) [inline README.md:43]: You should also mention that they need to have Firefox installed

File: app.py [hunk 5]
```diff
@@ -1,42 +1,44 @@
...
-@app.route('/items/<id>/todo')
-def move_item_todo(id):
-    trello_api.move_card_done_todo(id)
-    return redirect(url_for('index'))
+    return app
 
-
-if __name__ == '__main__':
-    app.run()
+app = create_app()
```
Comment (hugh-emerson) [inline app.py:41]: You do not want this here. When you run `flask run`, Flask will detect the create_app function and execute it.

Running it yourself here means that you are creating an instance of the app as soon as this file is imported and is not needed.

File: date_diff.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
+from datetime import datetime as dd, date
```
Comment (hugh-emerson) [inline date_diff.py]: This file is unused, right? So it can be deleted.

File: .gitignore [hunk 7]
```diff
@@ -116,4 +116,11 @@ env/
 .env
 
 ### Config.py file ###
-config.py
\ No newline at end of file
+config.py
+
+### geckodriver.exe and geckodriver.log ###
+geckodriver.exe
+geckodriver.log
```
Comment (hugh-emerson) [inline .gitignore:123]: Adding these to the .gitignore is correct, but it's not enough if the files are already being tracked. 
You need to `git rm --cached geckodriver.exe` etc as well.

File: test/test_trello_app.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    driver.get('http://localhost:5000/')
+    assert driver.title == 'To-Do App'
+    text_input = driver.find_element_by_xpath("//*[@id='NewItemTitle']")
+    text_input.send_keys("New Todo")    
+    driver.find_element_by_xpath("//button[contains(text(),'Add Item')]").click()    
+    driver.find_element_by_xpath("//a[contains(text(),'Mark as In Progress')]").click()
+    driver.find_element_by_xpath("//a[contains(text(),'Mark as Done')]").click()    
+    driver.find_element_by_xpath("//summary[contains(text(),'All Done Items')]").click()
+    driver.find_element_by_xpath("//a[contains(text(),'Mark as ToDo')]").click()
+
```
Comment (hugh-emerson) [inline test/test_trello_app.py:74]: You should have an assert at the end of the test to check that the final button behaved correctly (and possibly some asserts along the way)

File: test/test_app.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import pytest
```
Comment (hugh-emerson) [inline test/test_app.py:1]: pytest and selenium need to be added to your requirements.txt file in order to be able to run the tests without further manual setup.

File: README.md [hunk 10]
```diff
@@ -38,6 +38,17 @@ Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser
...
+
+test_view_model.py  - This is the unit test written as part of the TDD(Part1 Step 2 & 3) to add functionality to view model.
+test_trello_app.py - This is integration tests
+test_app.py - This is to test written to test end to end testing.
+
+
```
Comment (hugh-emerson) [inline README.md]: It would be nice to also mention how to run the tests, e.g. with `pytest` from command line.

**Review by hugh-emerson — APPROVED** [review]
The tests look good. The only major issue to fix is you haven't yet added pytest/selenium to requirements.txt.

Other than that I've left some minor suggestions but you're free to merge :+1:

File: .idea/DevOps-Course-Starter.iml (hunk 2 above)
Comment (shivaHanchate) [reply .idea/DevOps-Course-Starter.iml:5]: I was using PyCharm as my vscode was not working. Now the issue is resolved and I am only using vscode now.

File: README.md (hunk 4 above)
Comment (shivaHanchate) [reply README.md:43]: amended readme file

File: app.py (hunk 5 above)
Comment (shivaHanchate) [reply app.py:41]: I have removed app = create_app() as advised.

File: README.md (hunk 3 above)
Comment (shivaHanchate) [reply README.md]: amended readme file

File: date_diff.py (hunk 6 above)
Comment (shivaHanchate) [reply date_diff.py]: removed date_diff.py as not used.

File: .gitignore (hunk 7 above)
Comment (shivaHanchate) [reply .gitignore:123]: removed from cache as advised.

File: test/test_app.py (hunk 9 above)
Comment (shivaHanchate) [reply test/test_app.py:1]: added pytest and selenium to requirements.txt file

File: README.md (hunk 10 above)
Comment (shivaHanchate) [reply README.md]: As advised I have added pytest commands to run tests from command line

File: test/test_trello_app.py (hunk 8 above)
Comment (shivaHanchate) [reply test/test_trello_app.py:74]: added assert statements along the way until the end of the test.

**Pull request merged**